modules/transformation.py

```diff
@@ -0,0 +1,609 @@
+            )
+
+            # TODO using ReadAllFromText instead of ReadFromText[WithFilename] to avoid listing file sizes locally
+            #   https://stackoverflow.com/questions/60874942/avoid-recomputing-size-of-all-cloud-storage-files-in-beam-python-sdk
+            #   https://issues.apache.org/jira/browse/BEAM-9620
+            #   not an issue for the java SDK
+            #     https://beam.apache.org/releases/javadoc/2.37.0/org/apache/beam/sdk/io/contextualtextio/ContextualTextIO.Read.html#withHintMatchesManyFiles--
+            # files = p | beam.io.ReadFromTextWithFilename(matching)
+
+            files = (
+                p
+                # TODO replace with match continuously for streaming?
+                | fileio.MatchFiles(matching)
+                | "ExtractPath" >> beam.Map(lambda f: f.path)
+                | beam.Reshuffle()
+                | beam.io.ReadAllFromText(with_filename=True)
+            )
+
+        return files
+
+
+class WriteBigQuery(beam.PTransform):
+    def __init__(self, table, schema, streaming=None):
+        super().__init__()
+        self.table = table
+        self.schema = schema
+        self.streaming = streaming
+
+    def resolve_params(self):
+        if self.streaming:
+            # streaming pipeline
+            return {
+                "create_disposition": BigQueryDisposition.CREATE_IF_NEEDED,
+                "write_disposition": BigQueryDisposition.WRITE_APPEND,
+                "with_auto_sharding": True,
+                # parameters for STREAMING_INSERTS
+                "method": WriteToBigQuery.Method.STREAMING_INSERTS,
+                "ignore_unknown_columns": True,
+                "insert_retry_strategy": RetryStrategy.RETRY_ON_TRANSIENT_ERROR,
+                # parameters for FILE_LOADS
+                # "method": WriteToBigQuery.Method.FILE_LOADS,
+                # "triggering_frequency": 5 * 60,  # seconds
+                # "additional_bq_parameters": {"ignoreUnknownValues": True},
+            }
+        else:
+            # batch pipeline
+            return {
+                "create_disposition": BigQueryDisposition.CREATE_IF_NEEDED,
+                "write_disposition": BigQueryDisposition.WRITE_TRUNCATE,
+                "method": WriteToBigQuery.Method.FILE_LOADS,
+                "temp_file_format": FileFormat.JSON,
+                "additional_bq_parameters": {"ignoreUnknownValues": True},
+            }
+
+    def expand(self, pcoll, **kwargs):
+        return pcoll | WriteToBigQuery(
+            table=self.table, schema=self.schema, **self.resolve_params()
+        )
+
+
+def initialize_status_info(file_name, page):
+    # file name parsing kept for backward compatibility before 2022-03-01
+    dir_name, base_name = os.path.split(file_name)
+
+    # use crawl date instead of test date, should always be the first of the month
+    #  folder name contains the crawl date, file name contains the test date
+    # date = utils.test_date(page, base_name)
```

Comment:
This seems like more of a note to self, could it be removed now that it's fixed?

Reply:
Yep, will remove.